<p>Paralogue Annotation utilizes </p>

<p>The methods and evaluations are described fully in our manuscript available here [TODO: preprint-link]. The source code for the manuscript, along with all data and code necessary to reproduce the analyses, is available on <a href="https://github.com/ImperialCardioGenetics/Paralogue_Annotation_App">GitHub</a>.</p>

<p>The web app was built using <a href="http://shiny.rstudio.com">Shiny</a>, the source code is freely available at <a href="https://github.com/ImperialCardioGenetics/Paralogue_Annotation_App">https://github.com/ImperialCardioGenetics/Paralogue_Annotation_App</a>.</p>

<p>&nbsp;</p>

<u><h2>Recommended Usage</h2></u>
<!--<p>Although we showed the benefits of the proposed model for gene-disease classification and its superiority over existing genome-wide machine learning tools, we emphasize that CardioBoost is not intended to use as a standalone clinical decision tool to replace the whole ACMG guidelines <a href="https://www.ncbi.nlm.nih.gov/pubmed/25741868">(Richards et al. 2015)</a> for clinical variant interpretation. For example, a variant with higher than 90% pathogenicity score predicted by CardioBoost as pathogenic shouldnâ€™t straightforwardly be interpreted as Pathogenic without integrating other lines of evidences. Therefore, in the context of inherited cardiac conditions, the clinically-relevant classification by CardioBoost is intended to use as the evidence PP3 within ACMG guidelines as a more reliable and accurate computational tool over genome-wide ones in supporting variant interpretation in Cardiomyopathies and Arrhythmias.</p>-->
<br></br>

<br></br>
<u><h2>FAQ</h2></u>
<h3>Why does CardioBoost not output predictions on my input list of variants?</h3>
<p>There are mainly three reasons that CardioBoost would not return any prediction:</p>
<ul>
<li>The gene is not included as disease-related genes described above. Please check the gene lists above.</li>
<li>The mutation is not a valid missense change on the gene's canonical transcript (shown in the gene lists above).</li>
<li>The variant's gnomAD allele frequency is larger than 0.1%, which can be considered as a common variant and highly likely benign to cardiomyopathies and arrhythmias.</li>
</ul>
<br></br>


<u><h2>License</h2></u>
<p>The data provided here is available under the <a href="https://opendatacommons.org/licenses/odbl/summary/index.html">ODC Open Database License (ODbL) </a>: you are free to share and modify the data provided here as long as you attribute any public use of the database, or works produced from the database; keep the resulting data-sets open; and offer your shared or adapted version of the dataset under the same ODbL license.</p>

<p>The app is released under a GNU Lesser General Public License v2.1 [TODO:license-link].</p>
