<p>Paralogue Annotation utilizes information from evolutionarily related proteins, specifically paralogues, to help inform the clinical significance of missense variants associated with human diseases. The original methodology and implementation of <a href = "https://www.cardiodb.org/paralogue_annotation/">Paralogue Annotation on arrhythmia syndrome genes</a> was published <a href="https://onlinelibrary.wiley.com/doi/full/10.1002/humu.22114">here</a> and <a href = "https://jmg.bmj.com/content/51/1/35">here</a>. This web app extends Paralogue Annotation exome-wide, using paralogues defined by <a href="https://www.ensembl.org/Help/View?id=137"> Ensembl's gene trees</a> and pathogenic/likely pathogenic missense variants defined by <a href = "https://www.ncbi.nlm.nih.gov/clinvar/"> ClinVar</a>. </p>

<p>This web app is currently being built using <a href="http://shiny.rstudio.com">Shiny</a>, the source code is available at <a href="https://github.com/ImperialCardioGenetics/Paralogue_Annotation_App">https://github.com/ImperialCardioGenetics/Paralogue_Annotation_App</a>.</p>

<!--<p>Although we showed the benefits of the proposed model for gene-disease classification and its superiority over existing genome-wide machine learning tools, we emphasize that CardioBoost is not intended to use as a standalone clinical decision tool to replace the whole ACMG guidelines <a href="https://www.ncbi.nlm.nih.gov/pubmed/25741868">(Richards et al. 2015)</a> for clinical variant interpretation. For example, a variant with higher than 90% pathogenicity score predicted by CardioBoost as pathogenic shouldnâ€™t straightforwardly be interpreted as Pathogenic without integrating other lines of evidences. Therefore, in the context of inherited cardiac conditions, the clinically-relevant classification by CardioBoost is intended to use as the evidence PP3 within ACMG guidelines as a more reliable and accurate computational tool over genome-wide ones in supporting variant interpretation in Cardiomyopathies and Arrhythmias.</p>-->
<h2>Frequently Asked Questions (FAQ) </h2>
<p>
<b>Q. What genome build coordinates do my variants need to be in?</b><br>
<b>A.</b> Currently only GRCh37 coordinates are supported. We recommend using <a href="https://www.ensembl.org/Homo_sapiens/Tools/AssemblyConverter">Ensembl's liftover service</a> for coordinate conversions.<br>
<br>
<b>Q. What are the Para_z scores?</b><br>
<b>A.</b> The Para_z scores are a measure of paralogue conservation independently derived by <a href="https://genomemedicine.biomedcentral.com/articles/10.1186/s13073-020-00725-6"><i>Lal et al.</i> (2020)</a>. You may therefore find in your results that some Para_z scores do not agree with your expectations. This is because the paralogue alignments used to generate the scores are different to the alignments used here. The Para_z scores can thus be thought as a third-party confidence score of paralogue conservation across aligned positions.<br>
<br>
<b>Q. How can we use the conservation of Ref/Alt alleles to filter out results?</b><br>
<b>A.</b> That is not currently available in this version of the web app.<br>
<br>
<b>Q. What paralogue alignments do you use here?</b><br>
<b>A.</b> We utilize paralogue alignments at the protein level generated by Ensembl, which were obtained through <a href = "http://www.ensembl.org/info/docs/api/compara/compara_tutorial.html">Compara</a><br>
<br>
<b>Q. Why do the results for arrhythmia genes from the original <a href = "https://www.cardiodb.org/paralogue_annotation/">Paralogue Annotation</a> and here differ?</b><br>
<b>A.</b> This is mainly because the original Paralogue Annotation utilized T-COFFEE for the alignments, whereas Ensembl's alignments are generated by CLUSTAL W instead. Furthermore, variants from <a href = "http://www.hgmd.cf.ac.uk/ac/index.php">HGMD</a> were used instead of <a href = "https://www.ncbi.nlm.nih.gov/clinvar/"> ClinVar</a>.<br>
<br>
<b>Q. What formats do my input variants have to be in?</b><br>
<b>A.</b> Currently variants have to be submitted using their chromosome, position, reference allele, and alternate allele using any delimiter in the format of "CHROM:POS:REF:ALT" with separate variants on newlines. Alternatively we also accept <a href = "https://www.internationalgenome.org/wiki/Analysis/vcf4.0/">VCF</a> as well.<br>
<br>
<!--<b>Q. Can I query variants beyond Chromosome 1?</b><br>
<b>A.</b> This current DEMO version can only query variants in Chromosome 1. Future versions will be able to provide results for all chromosomes. Data for these are already available but have simply not been implemented here. If you wish to obtain all data for all chromosomes then please contact nyl112@ic.ac.uk
</p>-->
<br><br>
<p>For more details on specific methods, code of how Paralogue Annotation functions, or any other questions please email nyl112@ic.ac.uk</p>
<br><br>
<p>This web app is a <b>work in progress,</b> final version may differ.</p>
